File: test/aidbox_sdk/snapshots/dotnethl7-fhir-r4-core/ResearchElementDefinition.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.R4.Core;

public class ResearchElementDefinition : DomainResource, IResource
{
    public string? Description { get; set; }
    public string? Date { get; set; }
    public Base.ContactDetail[]? Endorser { get; set; }
    public string? Publisher { get; set; }
    public string? ApprovalDate { get; set; }
    public string? VariableType { get; set; }
    public Base.CodeableConcept[]? Jurisdiction { get; set; }
    public string? Purpose { get; set; }
    public Base.CodeableConcept? SubjectCodeableConcept { get; set; }
    public string? Name { get; set; }
    public Base.UsageContext[]? UseContext { get; set; }
    public string? Copyright { get; set; }
    public required string Type { get; set; }
    public bool? Experimental { get; set; }
    public Base.CodeableConcept[]? Topic { get; set; }
    public string? Title { get; set; }
    public string[]? Library { get; set; }
    public Base.ContactDetail[]? Author { get; set; }
    public required ResearchElementDefinition_Characteristic[] Characteristic { get; set; }
    public string? Usage { get; set; }
    public required string Status { get; set; }
    public string? Subtitle { get; set; }
    public string[]? Comment { get; set; }
    public string? Url { get; set; }
    public Base.Identifier[]? Identifier { get; set; }
    public string? LastReviewDate { get; set; }
    public Base.ContactDetail[]? Editor { get; set; }
    public Base.ContactDetail[]? Reviewer { get; set; }
    public string? ShortTitle { get; set; }
    public string? Version { get; set; }
    public Base.RelatedArtifact[]? RelatedArtifact { get; set; }
    public Base.ContactDetail[]? Contact { get; set; }
    public Base.ResourceReference? SubjectReference { get; set; }
    public object? Subject    
    {
        get
        {
            if (SubjectCodeableConcept is not null)
            {
                return SubjectCodeableConcept;
            }
    
            if (SubjectReference is not null)
            {
                return SubjectReference;
            }
    
            return null;
        }
    
        set
        {
            if (value?.GetType() == typeof(Base.CodeableConcept))
            {
                SubjectCodeableConcept = (Base.CodeableConcept)value;
                return;
            }
    
            if (value?.GetType() == typeof(Base.ResourceReference))
            {
                SubjectReference = (Base.ResourceReference)value;
                return;
            }
    
            throw new ArgumentException("Invalid type provided");
        }
    }
    public Base.Period? EffectivePeriod { get; set; }

    public class ResearchElementDefinition_Characteristic : BackboneElement
    {
        public object? Definition    
        {
            get
            {
                if (DefinitionExpression is not null)
                {
                    return DefinitionExpression;
                }
        
                if (DefinitionDataRequirement is not null)
                {
                    return DefinitionDataRequirement;
                }
        
                if (DefinitionCanonical is not null)
                {
                    return DefinitionCanonical;
                }
        
                if (DefinitionCodeableConcept is not null)
                {
                    return DefinitionCodeableConcept;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.ResourceExpression))
                {
                    DefinitionExpression = (Base.ResourceExpression)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.DataRequirement))
                {
                    DefinitionDataRequirement = (Base.DataRequirement)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    DefinitionCanonical = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.CodeableConcept))
                {
                    DefinitionCodeableConcept = (Base.CodeableConcept)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? StudyEffectiveTiming { get; set; }
        public bool? Exclude { get; set; }
        public Base.ResourceExpression? DefinitionExpression { get; set; }
        public string? ParticipantEffectiveDuration { get; set; }
        public string? StudyEffectiveDuration { get; set; }
        public Base.DataRequirement? DefinitionDataRequirement { get; set; }
        public string? DefinitionCanonical { get; set; }
        public string? StudyEffectiveGroupMeasure { get; set; }
        public string? ParticipantEffectiveTiming { get; set; }
        public string? ParticipantEffectiveGroupMeasure { get; set; }
        public string? StudyEffectiveDescription { get; set; }
        public string? ParticipantEffectiveDateTime { get; set; }
        public string? StudyEffectiveTimeFromStart { get; set; }
        public Base.CodeableConcept? UnitOfMeasure { get; set; }
        public Base.Period? ParticipantEffectivePeriod { get; set; }
        public object? StudyEffective    
        {
            get
            {
                if (StudyEffectiveTiming is not null)
                {
                    return StudyEffectiveTiming;
                }
        
                if (StudyEffectiveDuration is not null)
                {
                    return StudyEffectiveDuration;
                }
        
                if (StudyEffectivePeriod is not null)
                {
                    return StudyEffectivePeriod;
                }
        
                if (StudyEffectiveDateTime is not null)
                {
                    return StudyEffectiveDateTime;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(string))
                {
                    StudyEffectiveTiming = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    StudyEffectiveDuration = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.Period))
                {
                    StudyEffectivePeriod = (Base.Period)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    StudyEffectiveDateTime = (string)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public string? ParticipantEffectiveDescription { get; set; }
        public Base.CodeableConcept? DefinitionCodeableConcept { get; set; }
        public object? ParticipantEffective    
        {
            get
            {
                if (ParticipantEffectiveDuration is not null)
                {
                    return ParticipantEffectiveDuration;
                }
        
                if (ParticipantEffectiveTiming is not null)
                {
                    return ParticipantEffectiveTiming;
                }
        
                if (ParticipantEffectiveDateTime is not null)
                {
                    return ParticipantEffectiveDateTime;
                }
        
                if (ParticipantEffectivePeriod is not null)
                {
                    return ParticipantEffectivePeriod;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(string))
                {
                    ParticipantEffectiveDuration = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    ParticipantEffectiveTiming = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    ParticipantEffectiveDateTime = (string)value;
                    return;
                }
        
                if (value?.GetType() == typeof(Base.Period))
                {
                    ParticipantEffectivePeriod = (Base.Period)value;
                    return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.UsageContext[]? UsageContext { get; set; }
        public Base.Period? StudyEffectivePeriod { get; set; }
        public string? ParticipantEffectiveTimeFromStart { get; set; }
        public string? StudyEffectiveDateTime { get; set; }
    }
}

File: test/aidbox_sdk/snapshots/python/hl7_fhir_r4_core/ResearchElementDefinition.py
from typing import Optional, List
from dataclasses import dataclass, field
from base import UsageContext
from base import Period
from base import ContactDetail
from base import DataRequirement
from base import CodeableConcept
from base import Expression
from base import RelatedArtifact
from base import Timing
from base import Duration
from base import DomainResource
from base import Reference
from base import Identifier
from base import BackboneElement

@dataclass(kw_only=True)
class ResearchElementDefinitionCharacteristic(BackboneElement):
    study_effective_timing: Optional[Timing] = None
    exclude: Optional[bool] = None
    definition_expression: Optional[Expression] = None
    participant_effective_duration: Optional[Duration] = None
    study_effective_duration: Optional[Duration] = None
    definition_data_requirement: Optional[DataRequirement] = None
    definition_canonical: Optional[str] = None
    study_effective_group_measure: Optional[str] = None
    participant_effective_timing: Optional[Timing] = None
    participant_effective_group_measure: Optional[str] = None
    study_effective_description: Optional[str] = None
    participant_effective_date_time: Optional[str] = None
    study_effective_time_from_start: Optional[Duration] = None
    unit_of_measure: Optional[CodeableConcept] = None
    participant_effective_period: Optional[Period] = None
    participant_effective_description: Optional[str] = None
    definition_codeable_concept: Optional[CodeableConcept] = None
    usage_context: Optional[List[UsageContext]] = field(default_factory=list)
    study_effective_period: Optional[Period] = None
    participant_effective_time_from_start: Optional[Duration] = None
    study_effective_date_time: Optional[str] = None

@dataclass(kw_only=True)
class ResearchElementDefinition(DomainResource):
    description: Optional[str] = None
    date: Optional[str] = None
    endorser: Optional[List[ContactDetail]] = field(default_factory=list)
    publisher: Optional[str] = None
    approval_date: Optional[str] = None
    variable_type: Optional[str] = None
    jurisdiction: Optional[List[CodeableConcept]] = field(default_factory=list)
    purpose: Optional[str] = None
    subject_codeable_concept: Optional[CodeableConcept] = None
    name: Optional[str] = None
    use_context: Optional[List[UsageContext]] = field(default_factory=list)
    copyright: Optional[str] = None
    type: str
    experimental: Optional[bool] = None
    topic: Optional[List[CodeableConcept]] = field(default_factory=list)
    title: Optional[str] = None
    library: Optional[List[str]] = field(default_factory=list)
    author: Optional[List[ContactDetail]] = field(default_factory=list)
    characteristic: list[ResearchElementDefinitionCharacteristic] = field(default_factory=list)
    usage: Optional[str] = None
    status: str
    subtitle: Optional[str] = None
    comment: Optional[List[str]] = field(default_factory=list)
    url: Optional[str] = None
    identifier: Optional[List[Identifier]] = field(default_factory=list)
    last_review_date: Optional[str] = None
    editor: Optional[List[ContactDetail]] = field(default_factory=list)
    reviewer: Optional[List[ContactDetail]] = field(default_factory=list)
    short_title: Optional[str] = None
    version: Optional[str] = None
    related_artifact: Optional[List[RelatedArtifact]] = field(default_factory=list)
    contact: Optional[List[ContactDetail]] = field(default_factory=list)
    subject_reference: Optional[Reference] = None
    effective_period: Optional[Period] = None